### PR TITLE
Dead mans switch

### PR DESCRIPTION
**Pull request by KcPele**
## Description
Added challenge smart contract, test and description 
_Concise description of proposed changes, We recommend using screenshots and videos for better description_

## Additional Information

- [ ] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [ ] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

## Related Issues

_Closes #{issue number}_

_Note: If your changes are small and straightforward, you may skip the creation of an issue beforehand and remove this section. However, for medium-to-large changes, it is recommended to have an open issue for discussion and approval prior to submitting a pull request._

Your ENS/address:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Template For Challenge - ETH Tech Tree
```
Comment (escottalexander) [inline README.md]: ~~Template For Challenge~~ Replace with: Dead Man's Switch

File: README.md [hunk 2]
```diff
@@ -14,37 +16,50 @@ Before you begin, you need to install the following tools:
...
 
 Here are some helpful references:
-*Replace with real resource links if any*
+
 - [one](buidlguidl.com)
 - [two](buidlguidl.com)
 - [three](buidlguidl.com)
 
-*--End of challenge specific section--*
+_End of challenge specific section_
```
Comment (escottalexander) [inline README.md]: Remove this line. It was just there to help you know where to edit.

File: README.md [hunk 3]
```diff
@@ -14,37 +16,50 @@ Before you begin, you need to install the following tools:
...
+
+##### Objective:
+
+In this challenge, you will create a smart contract that implements a "Dead Man's Switch". This contract allows users to deposit funds (ETH or other tokens) and set a time interval for regular "check-ins". If the user fails to check in within the specified time frame, designated beneficiaries can withdraw the funds.
+
+##### Scenario:
+
+Imagine you want to ensure that your funds are securely passed on to your loved ones or trusted parties if something happens to you and you can no longer interact with the contract. This "Dead Man's Switch" will automate this process based on your predefined conditions.
 
 Here are some helpful references:
```
Comment (escottalexander) [inline README.md]: If you don't have any helpful resource links then remove this section.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        address[] memory beneficiaries = deadMansSwitch.getBeneficiaries(
+            THIS_CONTRACT
+        );
+        assertEq(beneficiaries.length, 1);
+        assertEq(beneficiaries[0], BENEFICIARY_1);
+    }
+
+    // // Test withdrawing funds by the user
+    function testWithdraw() public {
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:67]: This test is failing. There are two solutions. It is failing because when it tries to move the ether the testing contract doesn't have these methods:
``` solidity
    receive() external payable {}
    fallback() external payable {}
``` 
Without those methods attempts to send ether to a contract will fail. An alternative to adding those methods would be to use a EOA account that is not a contract. That would require adding these two lines to the top of the test so that it runs the test as another account that is not a contract.
```solidity
    vm.deal(NON_CONTRACT_USER, ONE_THOUSAND);
    vm.startPrank(NON_CONTRACT_USER);
```

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+     * @param beneficiary The address of the beneficiary.
+     * Requirements:
+     * - The beneficiary address must not be the zero address.
+     * - The beneficiary must not already be added.
+     * - Emits a `BeneficiaryAdded` event with the caller's address and the beneficiary address.
+     */
+    function addBeneficiary(address beneficiary) public {
+        require(beneficiary != address(0), "Invalid beneficiary address");
+        User storage user = users[msg.sender];
+        for (uint i = 0; i < user.beneficiaries.length; i++) {
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: For loops are strongly discouraged in Solidity because they lead to [issues](https://stackoverflow.com/questions/48113615/solidity-for-loop-over-a-huge-amount-of-data), just one example. When it is possible we should try to avoid them.

Lucky for us we can replace the beneficiaries array with a isBeneficiary mapping(address => bool) in the Users struct. Then all you need to do here is check if `user.isBeneficiary[beneficiary] == false`.

Then you can set them to true instead of adding to an array: user.isBeneficiary[beneficiary] = true;

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+     * - Emits a `Withdrawal` event with the beneficiary's address and the amount of Ether withdrawn.
+     */
+    function withdrawAsBeneficiary(address userAddress) public {
+        User storage user = users[userAddress];
+        require(
+            block.timestamp >= user.lastCheckIn + user.checkInInterval,
+            "Check-in interval has not passed"
+        );
+        bool isBeneficiary = false;
+        for (uint i = 0; i < user.beneficiaries.length; i++) {
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Here too we can avoid the for loop by utilizing the mapping:
```solidity
      require(user.isBeneficiary[msg.sender], "Caller is not a beneficiary");
```

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    ) public view returns (uint) {
+        return users[userAddress].checkInInterval;
+    }
+
+    /**
+     * @dev Gets the beneficiaries of the user.
+     * @param userAddress The address of the user.
+     * @return The beneficiaries of the user.
+     */
+    function getBeneficiaries(
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Since we are using a mapping now, this method will need to be changed to a helper method for getting the values of the mapping from the struct.
Something like this.
```solidity
function getUserBeneficiary(address userAddress, address beneficiary) public view returns (bool) {
      users[userAddress].isBeneficiary[beneficiary];
}
```
Make sure you update the NatSpec as well so that the user knows what to do.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    // Test the check-in functionality
+    function testCheckIn() public {
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+        deadMansSwitch.checkIn();
+        assertEq(deadMansSwitch.getLastCheckIn(THIS_CONTRACT), block.timestamp);
+    }
+
+    // Test adding a beneficiary
+    function testAddBeneficiary() public {
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:43]: You will need to update this test to use the new view method for isBeneficiary instead of the array.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        address userAddress
+    ) public view returns (address[] memory) {
+        return users[userAddress].beneficiaries;
+    }
+}
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol:149]: This also means it becomes negligible to add a removeBeneficiary method since all we have to do is set the value in the mapping to false. Let's add that in this file and don't forget some tests too.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+    }
+
+    // Test that the CheckInIntervalSet event is emitted correctly
+    function testEmitCheckInIntervalSetEvent() public {
+        vm.expectEmit(true, true, true, true);
+        emit DeadMansSwitch.CheckInIntervalSet(THIS_CONTRACT, INTERVAL);
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+    }
+}
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:184]: Don't forget tests for the new removeBeneficiary method I mentioned.

**Review by escottalexander — CHANGES_REQUESTED** [review]
Great job pulling this together! Left some suggestions for you. This is really close.

File: README.md [hunk 11]
```diff
@@ -1,4 +1,4 @@
-# Template For Challenge - ETH Tech Tree
+# Dead Man's Switch
```
Comment (escottalexander) [inline README.md]: Oh sorry, I should have been clearer. Can you change this to "Dead Man's Switch - ETH Tech Tree". This way it will match the other challenges.

File: README.md [hunk 12]
```diff
@@ -30,18 +30,12 @@ foundryup
 
 In this challenge, you will create a smart contract that implements a "Dead Man's Switch". This contract allows users to deposit funds (ETH or other tokens) and set a time interval for regular "check-ins". If the user fails to check in within the specified time frame, designated beneficiaries can withdraw the funds.
 
-##### Scenario:
```
Comment (escottalexander) [inline README.md]: Sorry again, KC. I meant to say you should remove the "Here are some helpful references: ..." section. You can add your scenario section back in. I liked it. 

The references section was just meant to be an example of where/how you could include some links to articles or other materials related to this challenge. If you don't know of any then you can remove the section.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 13]
```diff
@@ -3,18 +3,19 @@ pragma solidity >=0.8.0 <0.9.0;
...
+    event BeneficiaryAdded(address indexed user, address indexed beneficiary);
+    event BeneficiaryRemoved(address indexed user, address indexed beneficiary);
 
     // Struct to store user data
     struct User {
         uint balance;
         uint lastCheckIn;
         uint checkInInterval;
-        address[] beneficiaries;
+        mapping(address => bool) beneficiaries;
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Could we change the name of this to "isBeneficiary" or "beneficiaryLookup"? Because it is more of a lookup to see if a specific address is a beneficiary. I think that reads better when you are looking at where it is used and it helps the user understand it's purpose.

As it is right now, there is no way to return all beneficiaries from it and I feel like that is what it communicates if it is called beneficiaries.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 14]
```diff
@@ -159,13 +161,18 @@ contract DeadMansSwitch {
     }
 
     /**
-     * @dev Gets the beneficiaries of the user.
+     * @dev Gets to check if there is a beneficiary.
      * @param userAddress The address of the user.
-     * @return The beneficiaries of the user.
+     * @param beneficiary The address of the beneficiary.
+     * @return The check-in interval of the user.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Need to add requirements so the challenger knows what it needs to do.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 15]
```diff
@@ -159,13 +161,18 @@ contract DeadMansSwitch {
...
-        return users[userAddress].beneficiaries;
+    function isBeneficiary(
+        address userAddress,
+        address beneficiary
+    ) public view returns (bool) {
+        return users[userAddress].beneficiaries[beneficiary];
     }
+
+    receive() external payable {}
+    fallback() external payable {}
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: These methods are not needed so I would remove them. When I mentioned them in the last review I was trying to say that they would need to be added to the testing contract "DeadMansSwitch.t.sol" but only if you didn't add the NON_CONTRACT_USER deal and startPranks. Looks like you did so you should be good. Just remove them from here.

The risk in adding receive and fallback here is that it allows a user to deposit to the contract without the contract updating their specific balance. leading to funds stuck in the contract.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 16]
```diff
@@ -39,15 +39,30 @@ contract DeadMansSwitchTest is Test {
...
+        deadMansSwitch.removeBeneficiary(BENEFICIARY_1);
+        assertEq(
+            deadMansSwitch.isBeneficiary(THIS_CONTRACT, BENEFICIARY_1),
+            false
         );
-        assertEq(beneficiaries.length, 1);
-        assertEq(beneficiaries[0], BENEFICIARY_1);
     }
 
     // // Test withdrawing funds by the user
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol]: Remove the extra set of "//".

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        assertEq(deadMansSwitch.getBalance(THIS_CONTRACT), 0);
+    }
+
+    // Test that non-beneficiaries cannot withdraw funds
+    function testWithdrawAsNonBeneficiary() public {
+        deadMansSwitch.deposit{value: ONE_THOUSAND}();
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+        vm.warp(block.timestamp + INTERVAL + 1);
+        vm.startPrank(NON_CONTRACT_USER);
+        vm.expectRevert("Caller is not a beneficiary");
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol]: Change this line to `vm.expectRevert();`

Because we don't know what revert message to expect but we do know we want to it to revert in some way.

That way no matter what revert message the user chooses for this condition the test will still pass.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    // Test that beneficiaries cannot withdraw funds before the interval has passed
+    function testWithdrawAsBeneficiaryBeforeInterval() public {
+        deadMansSwitch.deposit{value: ONE_THOUSAND}();
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        vm.warp(block.timestamp + INTERVAL - 1);
+        vm.startPrank(BENEFICIARY_1);
+        vm.expectRevert("Check-in interval has not passed");
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol]: Change this line to `vm.expectRevert();`

Because we don't know what revert message to expect but we do know we want to it to revert in some way.

That way no matter what revert message the user chooses for this condition the test will still pass.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        vm.warp(block.timestamp + INTERVAL + 1);
+        vm.startPrank(BENEFICIARY_1);
+        deadMansSwitch.withdrawAsBeneficiary(THIS_CONTRACT);
+        assertEq(deadMansSwitch.getBalance(THIS_CONTRACT), 0);
+    }
+
+    // Test that non-beneficiaries cannot withdraw funds
+    function testWithdrawAsNonBeneficiary() public {
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:96]: Let's add another version of this test where a non-beneficiary tries to withdraw before the interval. It will be good to have that edge case covered just in case the challenger's logic has a hole.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+
+import "forge-std/Test.sol";
+import "../contracts/DeadMansSwitch.sol";
+
+contract DeadMansSwitchTest is Test {
+    DeadMansSwitch public deadMansSwitch;
+    address THIS_CONTRACT = address(this);
+    address NON_CONTRACT_USER = vm.addr(1);
+    address BENEFICIARY_1 = vm.addr(2);
+    uint ONE_THOUSAND = 10 wei;
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol]: Shouldn't ONE_THOUSAND be 1000 wei?

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+
+    // Test adding a beneficiary
+    function testAddBeneficiary() public {
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        assertEq(
+            deadMansSwitch.isBeneficiary(THIS_CONTRACT, BENEFICIARY_1),
+            true
+        );
+    }
+    // Test removing a beneficiary
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:51]: Add new line above to separate these properly.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+
+    // Test withdrawing funds by a beneficiary after the interval has passed
+    function testWithdrawAsBeneficiary() public {
+        deadMansSwitch.deposit{value: ONE_THOUSAND}();
+        deadMansSwitch.setCheckInInterval(INTERVAL);
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        vm.warp(block.timestamp + INTERVAL + 1);
+        vm.startPrank(BENEFICIARY_1);
+        deadMansSwitch.withdrawAsBeneficiary(THIS_CONTRACT);
+        assertEq(deadMansSwitch.getBalance(THIS_CONTRACT), 0);
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol]: Add an assertion that the beneficiaries balance was updated to the amount expected.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+        user.beneficiaries[beneficiary] = true;
+        emit BeneficiaryAdded(msg.sender, beneficiary);
+    }
+
+    /**
+     * @dev Removes a beneficiary for the user's funds.
+     * @param beneficiary The address of the beneficiary to remove.
+     * Requirements:
+     * - The beneficiary must be already added.
+     * - Emits a `BeneficiaryRemoved` event with the caller's address and the beneficiary address.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol:87]: Add a requirement that "The beneficiary is removed". Hopefully it is obvious to them but no harm in adding clarity.

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+    /**
+     * @dev Gets the check-in interval of the user.
+     * @param userAddress The address of the user.
+     * @return The check-in interval of the user.
+     */
+    function getCheckInInterval(
+        address userAddress
+    ) public view returns (uint) {
+        return users[userAddress].checkInInterval;
+    }
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Let's remove these methods completely: `getBalance`, `getLastCheckIn`, and `getCheckInInterval`. We can get the values of these in the tests with 
`(uint balance, uint lastCheckIn, uint checkInInterval) = deadmansswitch.users[ADDRESS];`
Or if you only need one of them you can do this:
```
(uint balance,,) = deadmansswitch.users[ADDRESS];
(,uint lastCheckIn,) = deadmansswitch.users[ADDRESS];
(,,uint checkInInterval) = deadmansswitch.users[ADDRESS];
```

Just no reason to require a user to build them if they aren't needed.

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    // Test adding a beneficiary
+    function testAddBeneficiary() public {
+        deadMansSwitch.addBeneficiary(BENEFICIARY_1);
+        assertEq(
+            deadMansSwitch.isBeneficiary(THIS_CONTRACT, BENEFICIARY_1),
+            true
+        );
+    }
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:49]: Because we are requiring that the beneficiary is not the zero address and that they can't already be a beneficiary, we should add tests to check that those cases are true.

**Review by escottalexander — CHANGES_REQUESTED** [review]
Hey KC, A few more changes suggested. Nice work implementing the new mapping. Looking great!

Feel free to reach out on Telegram if you have questions about my comments.

**Comment by KcPele** [comment]
Great

File: README.md (hunk 12 above)
Comment (KcPele) [reply README.md]: ok

File: packages/foundry/test/DeadMansSwitchTest.t.sol (hunk 20 above)
Comment (KcPele) [reply packages/foundry/test/DeadMansSwitchTest.t.sol]: my bad

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 26]
```diff
@@ -109,10 +110,10 @@ contract DeadMansSwitch {
     }
 
     /**
-     * @dev Allows beneficiaries to withdraw the user's funds if the check-in interval has passed.
+     * @dev Allows isBeneficiary to withdraw the user's funds if the check-in interval has passed.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: I think this was accidentally changed and should remain "beneficiaries"

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 27]
```diff
@@ -109,10 +110,10 @@ contract DeadMansSwitch {
...
      * @param userAddress The address of the user.
      * Requirements:
-     * - The caller must be one of the user's beneficiaries.
+     * - The caller must be one of the user's isBeneficiary.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: Same here. Should be "beneficiaries"

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 28]
```diff
@@ -123,56 +124,25 @@ contract DeadMansSwitch {
...
-    function getCheckInInterval(
-        address userAddress
-    ) public view returns (uint) {
-        return users[userAddress].checkInInterval;
-    }
-
     /**
      * @dev Gets to check if there is a beneficiary.
      * @param userAddress The address of the user.
      * @param beneficiary The address of the beneficiary.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol:138]: Should add "@return boolean representing whether the address is a beneficiary"

File: packages/foundry/contracts/DeadMansSwitch.sol [hunk 29]
```diff
@@ -123,56 +124,25 @@ contract DeadMansSwitch {
...
-     * @return The check-in interval of the user.
+     * Requirements:
+     * - The beneficiary must be added.
```
Comment (escottalexander) [inline packages/foundry/contracts/DeadMansSwitch.sol]: I would change this to "returns true if the provided beneficiary is a given userAddresses beneficiary, otherwise false"

File: packages/foundry/test/DeadMansSwitchTest.t.sol [hunk 30]
```diff
@@ -85,7 +98,16 @@ contract DeadMansSwitchTest is Test {
         deadMansSwitch.setCheckInInterval(INTERVAL);
         vm.warp(block.timestamp + INTERVAL + 1);
         vm.startPrank(NON_CONTRACT_USER);
-        vm.expectRevert("Caller is not a beneficiary");
+        vm.expectRevert();
+        deadMansSwitch.withdrawAsBeneficiary(THIS_CONTRACT);
+    }
+    // Test that non-beneficiaries cannot withdraw funds before the interval has passed
```
Comment (escottalexander) [inline packages/foundry/test/DeadMansSwitchTest.t.sol:105]: Add extra line above this so that spacing is the same throughout.

**Review by escottalexander — CHANGES_REQUESTED** [review]
Looking great! Nice work. Just a few small nit picks.

**Comment by KcPele** [comment]
Thanks for the feedback!

**Review by escottalexander — APPROVED** [review]
Nice work! Everything looks good.

**Comment by escottalexander** [comment]
Closed #22
